tiledb/sm/tile/tile_metadata_generator.cc

```diff
@@ -0,0 +1,571 @@
+    const Tile* tile, const Tile* tile_validity, const uint64_t cell_size) {
+  assert(tile != nullptr);
+
+  // Initialize to null.
+  void* min = nullptr;
+  void* max = nullptr;
+  uint64_t null_count = 0;
+
+  // Get pointers to the data and cell num.
+  auto value = (char*)tile->buffer()->data();
+  auto validity_values = static_cast<uint8_t*>(tile_validity->buffer()->data());
+  auto cell_num = tile->buffer()->size() / cell_size;
+
+  // Process cell by cell.
+  for (uint64_t c = 0; c < cell_num; c++) {
+    const bool is_null = validity_values[c] == 0;
+    min =
+        !is_null &&
+                (min == nullptr ||
+                 strncmp((const char*)min, (const char*)value, cell_size) > 0) ?
+            value :
+            min;
+    max =
+        !is_null &&
+                (max == nullptr ||
+                 strncmp((const char*)max, (const char*)value, cell_size) < 0) ?
+            value :
+            max;
+    value += cell_size;
+    null_count += is_null;
+  }
+
+  return {min, max, null_count};
+}
+
+template <>
+void TileMetadataGenerator::sum<int64_t>(int64_t* sum, const int64_t value) {
+  assert(sum != nullptr);
+
+  if (*sum > 0 && value > 0 &&
+      (*sum > std::numeric_limits<int64_t>::max() - value)) {
+    *sum = std::numeric_limits<int64_t>::max();
+    throw std::overflow_error("Overflow in sum computation");
+  }
+
+  if (*sum < 0 && value < 0 &&
+      (*sum < std::numeric_limits<int64_t>::min() - value)) {
+    *sum = std::numeric_limits<int64_t>::min();
+    throw std::overflow_error("Overflow in sum computation");
+  }
+
+  *sum += value;
+}
+
+template <>
+void TileMetadataGenerator::sum<uint64_t>(uint64_t* sum, uint64_t value) {
+  assert(sum != nullptr);
+
+  if (*sum > std::numeric_limits<uint64_t>::max() - value) {
+    *sum = std::numeric_limits<uint64_t>::max();
+    throw std::overflow_error("Overflow in sum computation");
+  }
+
+  *sum += value;
+}
+
+template <>
+void TileMetadataGenerator::sum<double>(double* sum, double value) {
+  assert(sum != nullptr);
+
+  if ((*sum < 0.0) == (value < 0.0) &&
+      std::abs(*sum) > std::numeric_limits<double>::max() - std::abs(value)) {
+    *sum = *sum < 0.0 ? std::numeric_limits<double>::lowest() :
+                        std::numeric_limits<double>::max();
+    throw std::overflow_error("Overflow in sum computation");
+  }
+
+  *sum += value;
+}
+
+template <class T, class SUM_T>
+ByteVec TileMetadataGenerator::sum(Tile* tile) {
+  assert(tile != nullptr);
+
+  // Zero sum.
+  ByteVec ret(8, 0);
+
+  // Get pointers to the data and cell num.
+  auto values = static_cast<T*>(tile->buffer()->data());
+  auto cell_num = tile->buffer()->size() / sizeof(T);
+  auto sum_data = (SUM_T*)ret.data();
+
+  // Process cell by cell, swallowing overflow exception.
+  try {
+    for (uint64_t c = 0; c < cell_num; c++) {
+      auto value = static_cast<SUM_T>(*static_cast<T*>(&values[c]));
+      sum<SUM_T>(sum_data, value);
```

Comment:
I don't think that's a good idea. I think we should push the `for` loop inside this function instead of calling the function in every iteration.

Reply:
@stavrospapadopoulos Are you trying to have vectorization here? Because the sum function already needs to do some checks that will prevent vectorization. The reason this is implemented this way is because c++ doesn't allow partial template specialization. I'm going to build a small program tomorrow to do performance testing. Then we can discuss adding one sum function per tiledb fixed data type.

Reply:
I ran some experiments and found that the performance difference is minimal, ie for 10GB data:
Time code: 1.74361
Time code stavros: 1.6865
So it could be argued that changing this is not worth it. But we can use partial template specialization for structs and the code is still clean (commit coming soon). This will also simplify greatly someone's potential attempt at vectorizing this.